File: Hard/815. Bus Routes/815. Bus Routes.py
from collections import defaultdict, deque
from typing import List

class Solution:
    def numBusesToDestination(self, routes: List[List[int]], source: int, target: int) -> int:
        if source == target:
            return 0

        stop_to_routes = defaultdict(list)

        for i, route in enumerate(routes):
            for stop in route:
                stop_to_routes[stop].append(i)

        visited_routes = set()
        queue = deque([(source, 0)])

        while queue:
            current_stop, steps = queue.popleft()

            for route_index in stop_to_routes[current_stop]:
                if route_index in visited_routes:
                    continue

                visited_routes.add(route_index)

                for next_stop in routes[route_index]:
                    if next_stop == target:
                        return steps + 1

                    queue.append((next_stop, steps + 1))

        return -1

File: Hard/815. Bus Routes/815. Bus Routes.ts
function numBusesToDestination(
    routes: number[][],
    source: number,
    target: number
): number {
    if (source === target) {
        return 0;
    }

    const stopToRoutes: Map<number, number[]> = new Map();

    for (let i = 0; i < routes.length; i++) {
        for (const stop of routes[i]) {
            if (!stopToRoutes.has(stop)) {
                stopToRoutes.set(stop, []);
            }
            stopToRoutes.get(stop)!.push(i);
        }
    }

    const visitedRoutes: Set<number> = new Set();
    const queue: [number, number][] = [[source, 0]];

    while (queue.length > 0) {
        const [currentStop, steps] = queue.shift()!;

        for (const routeIndex of stopToRoutes.get(currentStop) || []) {
            if (visitedRoutes.has(routeIndex)) {
                continue;
            }

            visitedRoutes.add(routeIndex);

            for (const nextStop of routes[routeIndex]) {
                if (nextStop === target) {
                    return steps + 1;
                }

                queue.push([nextStop, steps + 1]);
            }
        }
    }

    return -1;
}
